Remove parsing of command line arguments from `CliSettingsSource.__init__`.

trygve-baerland opened the pull request:
As part of the discussions in #610 and #654 , PR #611 needs to be reverted due to breaking backwards compatability. In particular, it stems from `CliSettingsSource.__init__` eagerly parsing command line arguments when passed with `cli_parse_args`.

This PR is meant as an alternative solution to the problem:

1. The call to `_load_env_vars` in `CliSettingsSource.__init__` is removed.
2. Checks for `CliSettingsSource.env_vars` put in places where the caller needs the arguments to have been parsed. As far as I found, this turned out to be during serialization (in `CliSettingsSource._serialized_args`), and during `CliSettingsSource.__call__`.
3. I've also added a test that's almost identical to the example found in #654 .

> NOTE: This PR is primarily meant as part of a discussion (and as a learning experience for me), so feel free to reject it on account of code quality or any other reason. Any feedback, though, is greatly appreciated.

hramezani commented:
Thanks @trygve-baerland for the PR.

@kschwab can you take a look at this PR?

trygve-baerland commented:
Thank you for your feedback, @kschwab. 

I think your approach makes a lot of sense. It also has the added benefit of being less prone to unintended side effects, which is what got us here in the first place :)

trygve-baerland commented:
@kschwab : I've made the changes as you suggested (you were bang on with it).

Further, I've rebased onto #655, and can wait with this PR until that's been merged.
I've also added back the test that will be removed by #655 .

trygve-baerland commented:
Super, @hramezani ! I've rebased onto `main`, so I think it's good for a second look, @kschwab

kschwab commented:
Looks good! Thanks @trygve-baerland for the changes.
